moonstreamapi/moonstreamapi/admin/migrations/add_selectors.py

```diff
@@ -0,0 +1,187 @@
+"""
+Add selectors to all moonworm tasks.
+"""
+import logging
+import json
+
+
+from bugout.exceptions import BugoutResponseException
+from web3 import Web3
+
+from ...settings import (
+    BUGOUT_REQUEST_TIMEOUT_SECONDS,
+    MOONSTREAM_ADMIN_ACCESS_TOKEN,
+    MOONSTREAM_MOONWORM_TASKS_JOURNAL,
+)
+from ...settings import bugout_client as bc
+from ...actions import get_all_entries_from_search
+
+logger = logging.getLogger(__name__)
+
+
+def fill_missing_selectors_in_moonworm_tasks() -> None:
+    """
+    Add selectors to all moonworm tasks.
+    """
+
+    batch_size = 100
+
+    moonworm_tasks = get_all_entries_from_search(
+        journal_id=MOONSTREAM_MOONWORM_TASKS_JOURNAL,
+        token=MOONSTREAM_ADMIN_ACCESS_TOKEN,
+        search_query="#task_type:moonworm !#version:2.0",
+        limit=batch_size,
+        content=True,
+    )
+
+    logger.info(f"Found {len(moonworm_tasks)} moonworm tasks versions 1.0")
+
+    entries_tags = []
+
+    ## batch tasks
+
+    for task_batch in [
+        moonworm_tasks[i : i + batch_size]
+        for i in range(0, len(moonworm_tasks), batch_size)
+    ]:
+        count = 0
+        for task in task_batch:
+            tags = ["version:2.0"]
+
+            ## get abi
+            try:
+                abi = json.loads(task.content)
+            except Exception as e:
+                logger.warn(
+                    f"Unable to parse abi from task: {task.entry_url.split()[-1]}: {e}"
+                )
+                continue
+
+            if "name" not in abi:
+                logger.warn(
+                    f"Unable to find abi name in task: {task.entry_url.split()[-1]}"
+                )
+                continue
+
+            if not any([tag.startswith("abi_selector:") for tag in task.tags]):
+                ## generate selector
+
+                abi_selector = Web3.keccak(
+                    text=abi["name"]
+                    + "("
+                    + ",".join(map(lambda x: x["type"], abi["inputs"]))
+                    + ")"
+                )[:4].hex()
+
+                tags.append(f"abi_selector:{abi_selector}")
+
+                count += 1
+
+            entries_tags.append(
+                {
+                    "entry_id": task.entry_url.split("/")[-1],  ## 😭
```

Comment:
Why we need entry id in tag?

Reply:
It request for https://github.com/bugout-dev/spire/blob/171b6eb0f8ce9656308a4112deab76af59f662ea/spire/journal/api.py#L1772 that endpoint which get entries_ids and tags as array.